### PR TITLE
[SPARK-40293][SQL] Make the V2 table error message more meaningful

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Add more info in the v2 table error message to make it more meaningful.

### Why are the changes needed?
When V2 catalog is not configured, Spark fails to access/create a table using the V2 API and silently falls back to attempting to do the same operation using the V1 Api. This happens frequently among the users. We want to have a better error message so that users can fix the configuration/usage issue by themselves.

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
existing tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 1]
```diff
@@ -844,10 +846,6 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
       s"Cannot delete from table ${table.name} where ${filters.mkString("[", ", ", "]")}")
   }
 
-  def deleteOnlySupportedWithV2TablesError(): Throwable = {
```
Comment (amaliujia) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:847]: Why did? Is it not longer used after this factoring?

Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:847]: We can use `operationOnlySupportedWithV2TableError(operation: String)` at L544

**Comment by huaxingao** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 2]
```diff
@@ -542,7 +542,9 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
   }
 
   def operationOnlySupportedWithV2TableError(operation: String): Throwable = {
-    new AnalysisException(s"$operation is only supported with v2 tables.")
+    new AnalysisException(s"$operation is only supported with v2 tables. To use" +
+      s" v2 tables, please config the catalog correctly using spark.sql.catalog" +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:546]: if catalog is not configured correctly, I think users will see table not found exception?

File: core/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -520,6 +520,11 @@
           "NATURAL CROSS JOIN."
         ]
       },
+      "OPERATION_ONLY_SUPPORTED_WITH_V2_TABLE" : {
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:523]: let's make it short as it's a sub-error-class. How about `TABLE_OPERATION`

File: core/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -520,6 +520,11 @@
...
+        "message" : [
+          "Table `spark_catalog`.`default`.`t1` does not support <operation>. Please check the current catalog and namespace to make sure the qualified table name is expected, and also check the catalog implementation which is configured by 'spark.sql.catalog.spark_catalog'."
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:525]: let's pass the real table name

Comment (cloud-fan) [reply core/src/main/resources/error/error-classes.json:525]: and also the catalog name

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: core/src/main/resources/error/error-classes.json (hunk 3 above)
Comment (huaxingao) [reply core/src/main/resources/error/error-classes.json:523]: Sounds good. Changed.

File: core/src/main/resources/error/error-classes.json (hunk 4 above)
Comment (huaxingao) [reply core/src/main/resources/error/error-classes.json:525]: Fixed. Thanks!

[the remark this answers is not in the document]
Comment (huaxingao) [reply core/src/main/resources/error/error-classes.json:525]: Fixed. Thanks!

File: core/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -520,6 +520,11 @@
           "NATURAL CROSS JOIN."
         ]
       },
+      "TABLE_OPERATION" : {
+        "message" : [
+          "Table '<catalog>'.'<nameSpace>'.'<tableName>' does not support <operation>. Please check the current catalog and namespace to make sure the qualified table name is expected, and also check the catalog implementation which is configured by \"spark.sql.catalog.spark_catalog\"."
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:525]:
```suggestion
          "Table <tableName> does not support <operation>. Please check the current catalog and namespace to make sure the qualified table name is expected, and also check the catalog implementation which is configured by \"spark.sql.catalog.<catalog>\"."
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 6]
```diff
@@ -541,8 +541,15 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
...
-    new AnalysisException(s"$operation is only supported with v2 tables.")
+  def operationOnlySupportedWithV2TableError(
+      catalog: String,
+      nameSpace: String,
+      tableName: String,
+      operation: String): Throwable = {
+    new AnalysisException(
+      errorClass = "UNSUPPORTED_FEATURE",
+      errorSubClass = "TABLE_OPERATION",
+      messageParameters = Array(catalog, nameSpace, tableName, operation))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:552]: for passing table name, we should pass `toSQLId(nameParts)`

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 7]
```diff
@@ -50,28 +50,44 @@ class ResolveSessionCatalog(val catalogManager: CatalogManager)
   override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
     case AddColumns(ResolvedV1TableIdentifier(ident), cols) =>
       cols.foreach { c =>
-        assertTopLevelColumn(c.name, "AlterTableAddColumnsCommand")
+        if (c.name.length > 1) {
+          throw QueryCompilationErrors.operationOnlySupportedWithV2TableError(
+            ident.catalog.getOrElse(""),
+            ident.database.getOrElse(""),
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:56]: `ident` must be qualified, we can do `ident.catalog.get` and `ident.database.get`

File: core/src/main/resources/error/error-classes.json (hunk 5 above)
Comment (huaxingao) [reply core/src/main/resources/error/error-classes.json:525]: Updated. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 6 above)
Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:552]: Fixed

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala (hunk 7 above)
Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:56]: Fixed. Thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 8]
```diff
@@ -541,8 +541,13 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
...
 
-  def operationOnlySupportedWithV2TableError(operation: String): Throwable = {
-    new AnalysisException(s"$operation is only supported with v2 tables.")
+  def operationOnlySupportedWithV2TableError(
+      nameParts: Seq[String],
+      operation: String): Throwable = {
+    new AnalysisException(
+      errorClass = "UNSUPPORTED_FEATURE",
+      errorSubClass = "TABLE_OPERATION",
+      messageParameters = Array(toSQLId(nameParts), operation))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:550]: The operation is a part of SQL statement, correct? Please, wrap it by `toSQLStmt()`.

Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala]: Fixed. Thanks!

**Comment by MaxGekk** [comment]
@huaxingao Could you fix PlanResolutionSuite and SparkThrowableSuite, please.

File: core/src/main/resources/error/error-classes.json [hunk 9]
```diff
@@ -565,6 +565,11 @@
           "<property> is a reserved table property, <msg>."
         ]
       },
+      "TABLE_OPERATION" : {
+        "message" : [
+          "Table <tableName> does not support <operation>. Please check the current catalog and namespace to make sure the qualified table name is expected, and also check the catalog implementation which is configured by \"spark.sql.catalog.spark_catalog\"."
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:570]:
```suggestion
          "Table <tableName> does not support <operation>. Please check the current catalog and namespace to make sure the qualified table name is expected, and also check the catalog implementation which is configured by \"spark.sql.catalog.<catalogName>\"."
```

Comment (cloud-fan) [reply core/src/main/resources/error/error-classes.json:570]: people can configure catalog implementation for every catalog.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 10]
```diff
@@ -541,8 +541,13 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
...
 
-  def operationOnlySupportedWithV2TableError(operation: String): Throwable = {
-    new AnalysisException(s"$operation is only supported with v2 tables.")
+  def operationOnlySupportedWithV2TableError(
+      nameParts: Seq[String],
+      operation: String): Throwable = {
+    new AnalysisException(
+      errorClass = "UNSUPPORTED_FEATURE",
+      errorSubClass = "TABLE_OPERATION",
+      messageParameters = Array(toSQLId(nameParts), toSQLStmt(operation)))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:550]: we may pass `ADD COLUMN with qualified column` as the operation. we shouldn't call `toSQLStmt` here.

**Review by cloud-fan — APPROVED** [review]

**Review by MaxGekk — CHANGES_REQUESTED** [review]

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @huaxingao and @cloud-fan @amaliujia for review.

**Comment by huaxingao** [comment]
Thanks a lot! @MaxGekk @cloud-fan @amaliujia

**Pull request closed without merging**